dlib/smart_pointers/observer_ptr.h

```diff
@@ -0,0 +1,62 @@
+#ifndef DLIB_OBSERVER_PTR_H
+#define DLIB_OBSERVER_PTR_H
+
+#include <utility>
+#include <type_traits>
+
+namespace dlib
+{
+    template<class T>
+    class observer_ptr
```

Comment:
I'm not super sold on the idea of `observer_ptr`.  Like why not `T*` for this kind of thing?  I've seen the arguments for it, but they don't really connect with me.

Reply:
For me it's the move semantics. The pointer is nulled after move.

Reply:
I think i accidentally slipped this in. I don't need it in this PR.